### PR TITLE
Turing SDK: Define and save Pyfunc Ensemblers

### PR DESCRIPTION
**Pull request by romanwozniak**
### Context:
This MR introduces a new `sdk` module, that contains a skeleton of Turing SDK, related to work with pyfunc ensemblers.

### Changes:
- Reformatted openapi specs (and broke down definitions into multiple fiels), to define a minimal spec, that contains only these API methods, that are used by SDK
- Generated REST client using [openapi-generator](https://github.com/OpenAPITools/openapi-generator) (hence, everything under [sdk/turing/generated](https://github.com/romanwozniak/turing/tree/sdk_save_ensembler/sdk/turing/generated) can be reviewed with less attention to details)
- Added `ProjectsController` into turing-api to exposed MLP projects to Turing SDK
- Moved PyFuncEnsembler interface from `engines/batch-ensembler` into `sdk` and added `sdk` as a dependency on `batch-ensembler`
- Defined SDK methods sufficient for listing projects, creating/updating/listing pyfunc ensemblers
- Added GitHub workflow for building/publishing SDK to pypi repo
- Added a [quickstart](https://github.com/romanwozniak/turing/tree/sdk_save_ensembler/sdk/samples/quickstart) example on how to use SDK

### To be done before merged:
- Unit tests
- Add Python docs where missing

### CODE REVIEW

File: .github/workflows/e2e.yaml [hunk 1]
```diff
@@ -22,6 +29,11 @@ jobs:
       with:
         go-version: "1.14"
 
+    - name: Reclaim some disk space
+      run: |
+        sudo apt-get clean
+        df -h
```
Comment (ashwinath) [inline .github/workflows/e2e.yaml:35]: Don't think this is required.

File: api/turing/it/database/database.go [hunk 2]
```diff
@@ -40,15 +40,15 @@ func migrate(db *sql.DB, dbName string) (*sql.DB, error) {
 
 	if migrations, err := gomigrate.NewWithDatabaseInstance(fmt.Sprintf("file://%s", migrationsFolder),
 		dbName, driver); err != nil {
-		return nil, err
+		return db, err
```
Comment (ashwinath) [inline api/turing/it/database/database.go:43]: What is the reason behind this change?

File: api/turing/middleware/authorization.go [hunk 3]
```diff
@@ -65,6 +66,32 @@ func (a *Authorizer) Middleware(next http.Handler) http.Handler {
 	})
 }
 
+func (a *Authorizer) FilterAuthorizedProjects(
+	user string,
+	projects []mlp.Project,
+	action string,
+) ([]mlp.Project, error) {
+	projectIDs := make([]string, 0)
+	var allowedProjects []mlp.Project
```
Comment (ashwinath) [inline api/turing/middleware/authorization.go:75]: Nitpick: I think moving this variable declaration just before using it improves readability. i.e moving this line to above 88.

File: sdk/tests/__init__.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+
+def json_serializer(o):
+    if isinstance(o, (date, datetime)):
+        return o.isoformat()
+    if isinstance(o, client.model_utils.ModelNormal):
+        return o.to_dict()
+
+
+project_1 = client.models.Project(
```
Comment (krithika369) [inline sdk/tests/__init__.py:12]: Curious - is this common practice to define test data in the `__init__.py` rather than pytest fixtures ?

File: api/turing/service/ensembler_service.go [hunk 5]
```diff
@@ -93,5 +98,5 @@ func (service *ensemblersService) Save(ensembler models.EnsemblerLike) (models.E
 	if err := result.Error; err != nil {
 		return nil, err
 	}
-	return ensembler, nil
+	return service.FindByID(ensembler.GetID(), EnsemblersFindByIDOptions{})
```
Comment (ashwinath) [inline api/turing/service/ensembler_service.go:101]: Why do you call FindByID again? Calling db.Save(ensembler) should mutate the ensembler value, shouldn't it?

Comment (ashwinath) [reply api/turing/service/ensembler_service.go:101]: Also, the error returned by this method will be confusing as a Save method might return a FindByID error.

**Comment by ashwinath** [comment]
I have reviewed the Go code for now, will review the SDK code on Thursday.

File: sdk/turing/_base_types.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        self._id = id
+        self._created_at = created_at
+        self._updated_at = updated_at
+
+    @property
+    def id(self) -> int:
+        return self._id
+
+    @id.setter
+    def id(self, id: int):
```
Comment (krithika369) [inline sdk/turing/_base_types.py:21]: Why do we have these setters? Do we expect to use them or is it for completeness?

**Review by krithika369 — APPROVED** [review]
It's a LARGE MR. 😅

Left some questions / minor comments but on the whole LGTM. Thanks!

File: sdk/turing/ensembler.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+        :param features: pandas.Series, containing a single row with input features
+        :param predictions: pandas.Series, containing a single row with all models predictions
+                `predictions['model-a']` will contain prediction results from the model-a
+        :param treatment_config: dictionary, representing the configuration of a treatment,
+                that should be applied to a given record. If the experiment engine is not configured
+                for this Batch experiment, then `treatment_config` will be `None`
+
+        :returns ensembling result (one of str, int, float, double or array)
+        """
```
Comment (ashwinath) [inline sdk/turing/ensembler.py:32]: missing a `pass` here?

File: sdk/turing/session.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        """
+        Update existing ensembler
+        """
+        return EnsemblerApi(self._api_client).update_ensembler(
+            project_id=ensembler.project_id,
+            ensembler_id=ensembler.id,
+            ensembler=ensembler)
+
+
+active_session: TuringSession = TuringSession(
```
Comment (ashwinath) [inline sdk/turing/session.py:138]: what's this for?

File: api/ensemblers.yaml [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Ensembler"
+
+  "/projects/{project_id}/ensemblers/{ensembler_id}":
+    get:
+      tags: [ "Ensembler" ]
+      operationId: "GetEnsemblerDetails"
+      summary: Returns Ensembler"s details by its ID
```
[the remark this answers is not in the document]
Comment (ashwinath) [reply api/ensemblers.yaml:63]: Hey it's an apostrophe!

File: sdk/Makefile [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+env:
+	pip install -r requirements.txt
+
+dev: env
+	pip install -r requirements.dev.txt
+
+gen-client:
+	rm -rf turing/generated
+	@openapi-generator generate \
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply sdk/Makefile:13]: Yeah, `README.md` is brewing 🍵

File: engines/batch-ensembler/ensembler/components/experimentation.py [hunk 11]
```diff
@@ -1,58 +0,0 @@
-from abc import abstractmethod
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply engines/batch-ensembler/ensembler/components/experimentation.py:1]: It was only temporarily added into `engines/batch-ensembler` until the SDK package is ready. It's expected from users to install `turing-sdk` to implement (and save) their pyfunc ensemblers, and not the `engines/`batch-ensembler` (which is just a spark application, that is executed inside spark cluster)

File: api/openapi-batch.yaml [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+      summary: Returns a list of Ensembling jobs
+      parameters:
+        - in: path
+          name: project_id
+          schema:
+            $ref: "#/components/schemas/Id"
+          required: true
+      responses:
+        "200":
+          description: TBU
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/openapi-batch.yaml:26]: Not exactly, it was already updated in `api/swagger-batch.yaml` by @ashwinath, but I'll clean it up in a separate PR this mess with openapi/swagger specs that we have now after I pulled @ashwinath's PR

File: api/openapi-sdk.yaml [hunk 13]
```diff
@@ -0,0 +1,13 @@
+openapi: 3.0.3
+info:
+  title: Turing Minimal Openapi Spec for SDK
+  version: 0.0.1
+servers:
+  - url: localhost:8080/v1
+paths:
+  "/projects":
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/openapi-sdk.yaml:8]: Likely, yes. In this case you wouldn't need "turing session", but rather can do everything within "mlp session", so the SDK can send calls directly to mlp api

File: sdk/samples/quickstart/quick_start.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    turing.set_url(turing_api)
+    turing.set_project(project)
+
+    # List projects
+    projects = turing.Project.list()
+    for p in projects:
+        print(p)
+
+    # Implement new pyfunc ensembler
+    class MyEnsembler(turing.ensembler.PyFunc):
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply sdk/samples/quickstart/quick_start.py:17]: Did it for the explicitness of the flow (i.e. to define "Implement new pyfunc ensembler" as a separate step in this process). But yeah, maybe it's not necessary

File: sdk/samples/quickstart/quick_start.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        ensembler_instance=MyEnsembler(),
+        conda_env={
+            'channels': ['defaults'],
+            'dependencies': [
+                'python=3.7.0',
+                "cookiecutter>=1.7.2",
+                "numpy"
+            ]
+        },
+        code_dir=["../samples"],
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply sdk/samples/quickstart/quick_start.py:64]: It doesn't matter. this code dir is just uploaded together with the pickled ensembler and added into the PYTHONPATH. So if the `code_dir` is defined `"../samples"` than it can be referred inside the ensembler code as `from samples.quickstart.quick_start import main`

File: sdk/tests/__init__.py (hunk 4 above)
Comment (romanwozniak) [reply sdk/tests/__init__.py:12]: It's wild wild west, everyone is doing it in their own way :)

but on a serious note, I didn't want to make it as individual fixtures (such as `project_1`, `project_2`...) because then they will have to be included into the test method's argument list and it could potentially grow arguments list unbearable long. 
I'll re-evaluate what can be made as a fixture after I'm done with all the tests

File: sdk/turing/_base_types.py (hunk 6 above)
Comment (romanwozniak) [reply sdk/turing/_base_types.py:24]: For completeness

File: .github/workflows/e2e.yaml (hunk 1 above)
Comment (romanwozniak) [reply .github/workflows/e2e.yaml:35]: yeah, maybe it's not. e2e test is still failing time to time even with this explicit clean-up step, so maybe it's unrelated

File: api/turing/it/database/database.go (hunk 2 above)
Comment (romanwozniak) [reply api/turing/it/database/database.go:43]: In some scenarios, it was possible that this snippet will throw nil pointer. For instance, if the migration fails (because of typo in the migration file), this line will return `nil` and then there will be a panic at https://github.com/gojek/turing/blob/b67c48b7e1b1777f89ec6c6860e3965c516fff9f/api/turing/it/database/database.go#L77, when `.Close()` is called on `nil`

File: api/turing/middleware/authorization.go (hunk 3 above)
Comment (romanwozniak) [reply api/turing/middleware/authorization.go]: agreed

File: sdk/turing/session.py (hunk 8 above)
Comment (romanwozniak) [reply sdk/turing/session.py:138]: To store the information about what project the user is working under and to provide the bridge between SDK methods, exposed to the user, and API endpoints

File: sdk/turing/ensembler.py (hunk 7 above)
Comment (romanwozniak) [reply sdk/turing/ensembler.py:32]: 💯

File: api/turing/service/ensembler_service.go (hunk 5 above)
Comment (romanwozniak) [reply api/turing/service/ensembler_service.go:101]: > Why do you call FindByID again? Calling db.Save(ensembler) should mutate the ensembler value, shouldn't it?

Only the fields, that were mutated explicitly (i.e. new `name` is set and then saved), but not for fields, updated automagically (`updated_at`).

Comment (romanwozniak) [reply api/turing/service/ensembler_service.go:101]: > Also, the error returned by this method will be confusing as a Save method might return a FindByID error.

That's less of the concern. It was done the same way for other services: https://github.com/gojek/turing/blob/b67c48b7e1b1777f89ec6c6860e3965c516fff9f/api/turing/service/router_service.go#L71

File: .github/workflows/e2e.yaml (hunk 1 above)
Comment (romanwozniak) [reply .github/workflows/e2e.yaml]: you are right, there is no any visible effect of this command, removed

File: api/turing/service/ensembler_service.go (hunk 5 above)
Comment (ashwinath) [reply api/turing/service/ensembler_service.go:101]: Hmm I wonder if gorm allows for the postgres keyword `RETURNING` though. But not consequential

File: sdk/turing/session.py (hunk 8 above)
Comment (ashwinath) [reply sdk/turing/session.py:138]: But what if the turing is not hosted on the user's machine? Importing this file will cause this to run and throw an exception?

File: api/turing/service/ensembler_service.go (hunk 5 above)
Comment (romanwozniak) [reply api/turing/service/ensembler_service.go:101]: No, gorm doesn't support it: https://github.com/go-gorm/gorm/issues/4088

File: sdk/turing/session.py (hunk 8 above)
Comment (romanwozniak) [reply sdk/turing/session.py:138]: It's just a default session, it's expected that user will initialize it with a proper Turing API before using:
```python
turing.set_url("http://turing.com/api", use_google_oauth=True)
```

If user has skipped this step, then the error will only occur when user will try to call Turing API with one of available methods, for instance `turing.Project.list()`. The error is not being thrown just on import.

**Review by ashwinath — APPROVED** [review]
Just some E2E tests to fix.

File: sdk/Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (romanwozniak) [reply sdk/Makefile:13]: ✅

File: sdk/tests/__init__.py (hunk 4 above)
Comment (romanwozniak) [reply sdk/tests/__init__.py]: This is addressed now too

**Comment by romanwozniak** [comment]
> Just some E2E tests to fix.

The test is flaky likely due to limited storage available in runner (but this theory has to be proved).
